### DISCUSSION TITLE
[Red-Knot] String Annotations

### DISCUSSION THREAD

**Discussion by Slyces** [astral-sh/ruff#13499] [General]
Hi, I am opening a discussion as I am not sure this belongs in a github issue.

As I worked on future annotations in red-knot a short while ago, I noticed that there is no support for string annotations yet. I did locate a couple `// TODO` related to this.

I might be interested in coding this feature at some point, however, I am not sure on how to proceed.

I know that given the string, I can use a the python parser to produce AST nodes and probably infer the type of a string annotation. However, this leads to problems in the current architecture as those AST nodes are not known for the SemanticIndex and symbols table.

Could we consider solving this by parsing StringAnnotations directly on the first run, making them into maybe a new variant? Or is there a way I don't see yet to achieve this by keeping those nodes as StringLiteral but parsing at inference time?

**Comment by MichaReiser** [comment]
Yeah, what you outline is correct. Stringified annotations are a pain, and we don't have a good solution yet to support them well. 

* Ruff parses the stringified annotations on demand. However, that comes with its own challenges: For example that all node offsets are off. This is more of a problem for Red Knot where we use offsets to uniquely identify nodes... 
* We could extend our AST and parse out the annotations. We have to consider that we want to use the same AST for the formatter so that type checking and formatting a file doesn't require parsing it twice. That means the AST needs to contain enough information for the formatter and type checker. The relevant information for the formatter is understanding the implicit concatenated string parts. So these should be retained in a representation.... However, it's valid to write `"Lit" "ereal[10]"` and it should be parsed as a single expression for type checking... We have to explore how an AST could support both use cases (e.g., introduce a new expression variant and store the string offsets separately?)

Whenever possible, I prefer to solve this in the parser using a new parser mode (so that we don't need to rewrite ruff). But it does require design on how to make this fit into the AST. 

I'm not sure if I remember it correctly but I believe @dhruvmanila wanted to look into this after this PTO but we can also support you designing and implementing this change if you're interested. 

CC: @carljm

**Reply by MichaReiser** [reply]
We should also look into what string kinds we have to support for string annotations. https://discuss.python.org/t/exotic-kinds-of-string-annotations/50527

**Reply by Slyces** [reply]
> We have to explore how an AST could support both use cases (e.g., introduce a new expression variant and store the string offsets separately?)

If you welcome experimentation draft PRs, I'd be happy to try to see what I can achieve with this direction

> but I believe @dhruvmanila wanted to look into this after this PTO but we can also support you designing and implementing this change if you're interested

If someone else is working on this, it's also fine as I do this on my free time I won't necessarily move super fast :P

**Reply by carljm** [reply]
Doing this in the parser and representing it in the AST as a distinct expression node kind seems pretty nice to me. I assume we want this to be an optional parser mode so that if someone only runs the formatter, we don't need to do the extra work of parsing stringified annotations.

We also need to handle partially stringified annotations, e.g. `SomeGenericType["ForwardReference"]`. So whatever handling we add here doesn't just apply to annotations that are strings at the top level, but to any string found within an annotation.

This gets particularly subtle because the parser can't reliably distinguish the above case from `Literal["foo"]`, where the `"foo"` really is just a string. (We can't just text match on the word `Literal` because we also have to handle aliasing, `from typing import Literal as L; x: L["foo"]`.) So even the `"foo"` in `Literal["foo"]` I think would need this handling in the parser, and it would only be in type inference (where we can resolve type of imports) that we would choose to handle `"foo"` in `Literal["foo"]` as a string, while we'd choose to handle `"ForwardReference"` in `SomeGenericType["MyForwardReference"]` as a deferred expression.

**Reply by MichaReiser** [reply]
> This gets particularly subtle because the parser can't reliably distinguish t

Isn't it sufficient for the parser to know that it is currently inside of a type annotation position or are there cases where an expression can be a stringified type or a regular expression?

>  I assume we want this to be an optional parser mode so that if someone only runs the formatter, we don't need to do the extra work of parsing stringified annotations.

My main motivation for a parser mode is so that we don't need to rewrite Ruff linter and formatter just yet. The long term goal should be that both the formatter and type checker use the same AST (and parse mode) because it would otherwise be necessary to parse and cache the same file twice.

**Reply by carljm** [reply]
> Isn't it sufficient for the parser to know that it is currently inside of a type annotation position

Yes, I think that is sufficient; in the parser, we should apply this handling to any string found in annotation position or as a sub-expression of an expression in annotation position.

> or are there cases where an expression can be a stringified type or a regular expression?

I think the case I described above is such a case: in `Literal["foo"]` we can have a string inside an annotation expression, which is just a string, it's not a stringified type. But we can't determine this reliably in the parser, we can only determine it in type inference. So it's important that our handling of strings in annotation position makes available both the original string and the parsed expression (I think we need this for the formatter anyway?) so that type inference can decide to treat it as a normal string when it's the subscript to `typing.Literal`, and use the parsed expression otherwise.



> The long term goal should be that both the formatter and type checker use the same AST (and parse mode) because it would otherwise be necessary to parse and cache the same file twice.

Makes sense. I guess then we just pay a bit of extra cost if someone only runs the formatter, but that's OK.

**Reply by MichaReiser** [reply]
> makes available both the original string and the parsed expression (

This is something I want to evaluate carefully. I would rather prefer if the node only has a single representation. We already have an `AnyString` abstraction over all string-like types and we could use it here too. I'm not saying that this is definitely possible. 

Overall, we should explore:

* Ruff's current approach
* A single node that represents parsed type annotations
* A "dual" node representing both the string literal and the parsed type annotation

**Reply by carljm** [reply]
Sure, I think more evaluation makes sense. But I don't think "A single node that represents parsed stringified type annotations" is possible, if we do this in the parser. Type inference has to have some way to recover either the original string or its parsed version, otherwise we can't handle `Literal` correctly.

**Reply by MichaReiser** [reply]
The week before, I noticed that Ruff doesn't correctly format stringified type annotations unlike Black https://github.com/astral-sh/ruff/issues/13479. I haven't looked into exactly how Black formats stringified type annotations but the formatter might not need the full string representation (unless Black preserves the existing strings). It might only be the linter that requires the "full" string literals. 

What's unclear is how error recovery and error-reporting should work. Let's say a type annotation has the value `X[ABC` where the closing `]` is missing. Should the parser try to recover to `X[ABC]` or parse the string literal as is, considering that it is valid syntax? Should the parser emit a diagnostic or not? If not, which part is responsible for emitting a diagnostic?

**Reply by carljm** [reply]
> It might only be the linter that requires the "full" string literals.

I'm confused by this, because I mentioned already that the type checker will also need to be able to get access to the full original string literal in some cases. 

> What's unclear is how error recovery and error-reporting should work. Let's say a type annotation has the value `X[ABC` where the closing `]` is missing. Should the parser try to recover to `X[ABC]` or parse the string literal as is, considering that it is valid syntax?

If this string is in annotation position, and we are parsing stringified annotations in the parser and providing an AST for the string contents to consumers of the parser, then in this case we should still provide an AST with an invalid-syntax node, we shouldn't just provide a string. If we are type checking this code, we should emit a diagnostic on this. But if we are just linting (or providing language services) for the code, without type checking turned on, we should probably not emit a diagnostic for this invalid stringified annotation.

> Should the parser emit a diagnostic or not? If not, which part is responsible for emitting a diagnostic?

Does the parser ever emit diagnostics? I thought it just created invalid-syntax nodes and recovered, and it's up to the consumer to decide how to handle those nodes. That's how I think it should work, I don't think the parser itself should emit diagnostics.

**Reply by MichaReiser** [reply]
> I'm confused by this, because I mentioned already that the type checker will also need to be able to get access to the full original string literal in some cases.

Sorry I missed that

> We also need to handle partially stringified annotations, e.g. SomeGenericType["ForwardReference"]. So whatever handling we add here doesn't just apply to annotations that are strings at the top level, but to any string found within an annotation.

Considering that, handling it in the parser seems less appealing to me because the result is that some string annotations are handled in the parser while others still require special handling in more higher level crates. 

> Does the parser ever emit diagnostics? 

The parser always emits a diagnostic when it recovers.

**Reply by carljm** [reply]
> Considering that, handling it in the parser seems less appealing to me because the result is that some string annotations are handled in the parser while others still require special handling in more higher level crates.

I would assume that if we handle it in the parser, the parser would track recursively when it is within an annotation, and provide this handling for all strings found within an annotation.

I agree that if we can't (or don't want to) do that, then we shouldn't do this in the parser at all.

**Reply by MichaReiser** [reply]
> I would assume that if we handle it in the parser, the parser would track recursively when it is within an annotation, and provide this handling for all strings found within an annotation.

The issue I see is that some strings are indeed just strings and there's no good way for the parser to know which string is a string and which isn't. It leads to very obscure results if you want the recovery logic described here

> If this string is in annotation position, and we are parsing stringified annotations in the parser and providing an AST for the string contents to consumers of the parser, then in this case we should still provide an AST with an invalid-syntax node

because the parser then tries to parse the inner string of `Literal["TestString with spaces"]` as an expression, where it probably has to parse out a tuple or three identifiers `(TestString, with spaces)` as the "closest" valid expression, which is very far-fetched and unnecessary.

**Reply by carljm** [reply]
If we don't want to ever speculatively parse a string in an annotation that probably is (but might in the end not be) a stringified annotation, then I agree we can't do this in the parser.

It will also mean some difficulty for red-knot, since a lot of our architecture currently assumes that we have the full AST up-front, and we can number nodes, semantic index, etc. But I suspect we can handle this by treating each stringified annotation as almost a separate "file", but with some additional prepopulated builtin/global names, or something along those lines. Name lookup in stringified annotations is going to be weird and special-cased anyway (e.g. see the issue @Viicos mentioned below.)

**Reply by MichaReiser** [reply]
> It will also mean some difficulty for red-knot, since a lot of our architecture currently assumes that we have the full AST up-front, and we can number nodes, 

Possibly yes, unless we can also number the parsed out type annotations when we number the main AST.

**Comment by Viicos** [comment]
On a slightly related note, is red knot going to resolve forward annotations the same way it does for "normal" annotations? Because of a backward compatibility behavior of `typing.get_type_hints`, module scope is given priority for forward annotations (see https://github.com/microsoft/pyright/discussions/8958).

**Reply by carljm** [reply]
We haven't discussed this explicitly yet, but I suspect we will have to match the behavior of `typing.get_type_hints` and pyright here.

**Reply by Viicos** [reply]
Makes sense, thanks for the answer (and anyway hopefully string annotations will be a thing of the past at some point with PEP 649)

**Comment by carljm** [comment]
We do need to implement support for stringified annotations, which means we should have an open issue for it. So I am going to close this discussion for now and open an issue for implementing stringified annotations, with summaries of some of the things discussed here, and a reference back to this discussion.

**Comment by carljm** [comment]
https://github.com/astral-sh/ruff/issues/13796
